src/compiler/checker.ts

```diff
@@ -18056,7 +18278,30 @@ namespace ts {
         }
 
         function getLengthOfTupleType(type: TupleTypeReference) {
-            return getTypeReferenceArity(type) - (type.target.hasRestElement ? 1 : 0);
+            return type.target.fixedLength;
+        }
+
+        function getEndLengthOfTupleType(type: TupleTypeReference) {
+            return getTypeReferenceArity(type) - findLastIndex(type.target.elementFlags, f => !!(f & ElementFlags.Variable)) - 1;
+        }
+
+        function getElementTypeOfTupleType(type: TupleTypeReference, index: number, endSkipCount = 0) {
+            const length = getTypeReferenceArity(type) - endSkipCount;
+            if (index < length) {
+                const typeArguments = getTypeArguments(type);
+                const elementTypes: Type[] = [];
+                for (let i = index; i < length; i++) {
+                    const t = typeArguments[i];
+                    elementTypes.push(type.target.elementFlags[i] & ElementFlags.Variadic ? getIndexedAccessType(t, numberType) : t);
+                }
+                return getUnionType(elementTypes);
+            }
+            return undefined;
+        }
+
+        function isMatchingTupleTypes(t1: TupleTypeReference, t2: TupleTypeReference) {

```

Comment:
This doesn't have the very similar semantics to other `isMatching` functions we have; I think a comment is warranted or maybe a name change. I was expecting this function to compare the types themselves, not just the overall "binding shape" if you will

Reply:
Maybe `isTupleTypeStructureMatching`?